extension/encoding/encoding.go

```diff
@@ -0,0 +1,34 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package encoding // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding"
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+var errNoExtensionFound = "no extension found: %s"
+var errNotAnEncodingExtension = "extension %q doesn't implement %T"
+
+/*
+Receivers wanting to Unmarshal can use the GetEncoding like following:
+
+	unmarshaler, err := GetEncoding[(pmetric|ptrace|plog).Unmarshaler](host.GetExtensions(), id)
+
+Exporters wanting to Marshal can use the GetEncoding like following:
+
+	marshaler, err := GetEncoding[(pmetric|ptrace|plog).Marshaler](host.GetExtensions(), id)
+*/
+func GetEncoding[T any](extensions map[component.ID]component.Component, componentID component.ID) (T, error) {
```

Comment:
I understand you want to reduce code duplication, but this API doesn't seem intuitive.
- It's hard to understand what it's doing without reading the comments.
- It's called `encoding.GetEncoding`, but it can be applicable to any extension type. I can use it as `encoding.GetEncoding[storage.Extension]` as well. 
- It's not that much of a logic to put into a separate function. It reduces the code but doesn't improve readability.

I'd suggest not introducing this helper and keeping this code in all the exporters/receivers. Even if it brings some mode duplicated code, I think it makes it easier to understand.

Reply:
Potentially, we can introduce with helper in core, but let's hold off for now